indexing-service/src/main/java/org/apache/druid/indexing/overlord/hrtr/HttpRemoteTaskRunner.java

```diff
@@ -1197,109 +1159,82 @@ private void startPendingTaskHandling()
-            // And, it is automatically removed by any of the task execution threads when they notice that
-            // ti.getState().isPending() is false (at the beginning of this loop)
-            ti.setState(HttpRemoteTaskRunnerWorkItem.State.PENDING_WORKER_ASSIGN);
-            taskItem = ti;
-            break;
-          }
-
-          if (taskItem == null) {
-            // Either no pending task is found or no suitable worker is found for any of the pending tasks.
-            // statusLock.notifyAll() is called whenever a new task shows up or if there is a possibility for a task
-            // to successfully get worker to run, for example when a new worker shows up, a task slot opens up
-            // because some task completed etc.
-            statusLock.wait(TimeUnit.MINUTES.toMillis(1));
+          // Exhausted worker assignment retries, let's backoff a bit
+          if (workerToAssign == null) {
+            log.warn("Failed to find workers available to run task[%s]. Sending to back of queue", taskId);
+            pendingTasks.put(taskItem.withFreshSequenceNumber());
             continue;
           }
-        }
 
-        String taskId = taskItem.getTaskId();
+          // Mark this worker as unassignable while task is being assigned
+          workers.compute(
+              workerToAssign.getWorker().getHost(), (key, entry) -> {
+                Preconditions.checkState(
+                    entry != null,
+                    "Expected selected worker[%s] to be available",
+                    entry.getWorker().getHost()
+                );
+                Preconditions.checkState(
+                    entry.getState() == WorkerHolder.State.READY,
+                    "Expected worker[%s] state to be READY, got [%s]",
+                    entry.getWorker().getHost(),
+                    entry.getState()
+                );
 
-        if (taskItem.getTask() == null) {
-          log.makeAlert("No Task obj found in TaskItem for taskID[%s]. Failed.", taskId).emit();
-          // taskComplete(..) must be called outside of statusLock, see comments on method.
-          taskComplete(
-              taskItem,
-              null,
-              TaskStatus.failure(
-                  taskId,
-                  "No payload found for this task. "
-                  + "See overlord logs and middleManager/indexer logs for more details."
-              )
+                entry.setState(WorkerHolder.State.PENDING_ASSIGN);
+                return entry;
+              }
           );
-          continue;
-        }
 
-        if (immutableWorker == null) {
-          throw new ISE("Unexpected state: null immutableWorker");
+          // Mark this task as pending worker assign
+          tasks.compute(
+              taskId,
+              (key, entry) -> {
+                Preconditions.checkState(entry != null, "Expected task[%s] to be in tasks set", taskId);
+                Preconditions.checkState(
```

Comment:
Can we just log an error / raise an alert and continue with the loop?
That's what would happen anyway. It feels weird to throw these exceptions inside the lambda and have the outermost catch process them as generic unknown exceptions.